### PR TITLE
Merge pull request #4434 from ChengJin01/riscv_omr_part4_port_dir

### PR DESCRIPTION
**Pull request by ChengJin01**
The changes mainly include code changes at the
port directory to enable RISC-V 64bit
from the OMR perspective.

Issue: eclipse#4426

Signed-off-by: Cheng Jin <jincheng@ca.ibm.com>

### CODE REVIEW

**Comment by ChengJin01** [comment]
Already verified the compilation on Fedora_Linux/QEMU.

Reviewer: @keithc-ca , @pshipton , @DanHeidinga

**Comment by janvrany** [comment]
I'd like to suggest few improvements to thiis PR, if I may: 

 1. We need to define `OMRPORT_ARCH_RISCV `, see [f2ee53e2](https://github.com/janvrany/omr/commit/f2ee53e25f44d70ea5f01f17162bc7e1053608e8)
 2. Provide a basic implementation of `omrsysinfo_get_processor_description()`, see [95282e99](https://github.com/janvrany/omr/commit/95282e99988ba6452af3ac258353ec3ec1537f31)

With this PR rebased atop of current master and with above fixes, all OMR port test pass (I mean, all tests in `fvtest/port/omrporttest`).

**Comment by ChengJin01** [comment]
@janvrany,

> We need to define OMRPORT_ARCH_RISCV , see f2ee53e2

OMRPORT_ARCH_RISCV was already defined at include_core/omrport.h in https://github.com/eclipse/omr/pull/4435. I will move the change to this PR as it is more suitable to be here.

> Provide a basic implementation of omrsysinfo_get_processor_description(), see 95282e99

I need to compile a cross-build with your changes to see whether it works good before including it in this PR.

File: port/linuxriscv64/omrsignal_context.c [hunk 1]
```diff
@@ -73,11 +62,8 @@ infoForSignal(struct OMRPortLibrary *portLibrary, struct OMRUnixSignalInfo *info
 	case OMRPORT_SIG_SIGNAL_CODE:
 	case 3:
 		*name = "Signal_Code";
-		if( NULL != info->sigInfo) {
-			*value = &info->sigInfo->si_code;
-			return OMRPORT_SIG_VALUE_32;
-		}
-		break;
+		*value = &info->sigInfo->si_code;
```
Comment (DanHeidinga) [inline port/linuxriscv64/omrsignal_context.c]: Why are the null checks not needed?  What do other platforms do?

File: port/linuxriscv64/omrsignal_context.c [hunk 2]
```diff
@@ -89,75 +75,195 @@ infoForSignal(struct OMRPortLibrary *portLibrary, struct OMRUnixSignalInfo *info
...
-void*
-walkReserveSpace(uint32_t magic_to_find, uint8_t *reserved_space)
-{
-	Assert_PRT_true(0 && "Not yet implemented");
-	return NULL;
+	if ((index >= 0) && (index < NFPRS)) {
+		*name = n_fpr[index];
+		*value = &(context->fpregs[index]);
+		return OMRPORT_SIG_VALUE_FLOAT_64;
+	} else {
```
Comment (DanHeidinga) [inline port/linuxriscv64/omrsignal_context.c]: nit pick - the `else {` isn't required as the `if` block returns.

Clearer to write:
```c
if ((index >= 0) && (index < NFPRS)) {
	.....
	.....
	return OMRPORT_SIG_VALUE_FLOAT_64;
}
return OMRPORT_SIG_VALUE_UNDEFINED;
```
as it becomes obvious there is no other paths to the end of the method

File: port/linuxriscv64/omrsignal_context.c [hunk 3]
```diff
@@ -89,75 +75,195 @@ infoForSignal(struct OMRPortLibrary *portLibrary, struct OMRUnixSignalInfo *info
...
-{
-	Assert_PRT_true(0 && "Not yet implemented");
-	return 0;
+	*name = "";
+
+	if ((index >= 0) && (index < NGPRS)) {
+		*name = n_gpr[index];
+		*value = &(context->gregs[index]);
+		return OMRPORT_SIG_VALUE_ADDRESS;
+	} else {
```
Comment (DanHeidinga) [inline port/linuxriscv64/omrsignal_context.c]: Same nitpick as above

**Comment by ChengJin01** [comment]
All changes suggested by @janvrany has been included in this PR and will compile a cross-build to ensure everything works good in compilation.

File: port/linuxriscv64/omrsignal_context.c (hunk 2 above)
Comment (ChengJin01) [reply port/linuxriscv64/omrsignal_context.c]: Updated the code as suggested above.

File: port/linuxriscv64/omrsignal_context.c (hunk 3 above)
Comment (ChengJin01) [reply port/linuxriscv64/omrsignal_context.c]: Updated the code as suggested above.

File: port/linuxriscv64/omrsignal_context.c (hunk 1 above)
Comment (ChengJin01) [reply port/linuxriscv64/omrsignal_context.c]: In theory we can add these null checks if the passed-in `gpinfo ` is null (actually it shouldn't be null once the signal is triggered)
e.g.
```
runtime\vm\gphandle.c
vmSignalHandler(struct J9PortLibrary* portLibrary, U_32 gpType, void* gpInfo, void* userData)
{
...
	crashData.gpInfo = gpInfo;
...
	j9sig_protect(
		writeCrashDataToConsole, &crashData,
		recursiveCrashHandler, &recursiveCrashData,
		J9PORT_SIG_FLAG_MAY_RETURN | J9PORT_SIG_FLAG_SIGALLSYNC,
		&result);
```
Beside, I didn't see the null check is added on any other platform in the existing implementation.

Comment (DanHeidinga) [reply port/linuxriscv64/omrsignal_context.c]: Ok.  I'm fine with leaving it out then

**Comment by ChengJin01** [comment]
The cross-compilation passed with the latest update and verified on Fedora_riscv:
```
[jincheng@stage4 RISCV_OPENJ9]$ jdk11_rv64_cross_01_16_2020/bin/java -version
openjdk version "11.0.6-internal" 2020-01-14
OpenJDK Runtime Environment (build 11.0.6-internal+0-adhoc.root.openj9-openjdk-jdk11)
Eclipse OpenJ9 VM (build riscv_openj9_v2_uma-aed689f, JRE 11 Linux riscv-64-Bit Compressed References 20200116_000000 (JIT disabled, AOT disabled)
OpenJ9   - aed689f
OMR      - e499106
JCL      - ea635f3 based on jdk-11.0.6+4)
```

**Comment by DanHeidinga** [comment]
@rwy0717 Can you review this and launch the necessary builds?

**Comment by rwy0717** [comment]
@genie-omr build all

**Pull request merged**